src/libclipper/test/policy_test.cpp

```diff
@@ -0,0 +1,261 @@
+  VersionedModelId model_3 = std::make_pair("random_forest", 2);
+  int times = 1000;
+  PolicyState state;
+};
+
+TEST_F(Exp4Test, UpdateTest) {
+
+  auto feedback = Utility::create_feedback(20);
+  std::vector<Output> predictions;
+  while (times > 0) {
+    auto y_hat = rand() % 100;
+    if (y_hat < 33) {
+      predictions = Utility::create_predictions(model_1, y_hat);
+    } else if (y_hat > 66) {
+      predictions = Utility::create_predictions(model_3, y_hat);
+    } else {
+      predictions = Utility::create_predictions(model_2, y_hat);
+    }
+    state = Exp4Policy::process_feedback(state, feedback, predictions);
+    times -= 1;
+  }
+  ASSERT_GT(state.model_map_[model_1]["weight"],
+            state.model_map_[model_2]["weight"]);
+  ASSERT_GT(state.model_map_[model_2]["weight"],
+            state.model_map_[model_3]["weight"]);
+}
+
+TEST_F(Exp4Test, SelectionTest) {
+  auto query = Utility::create_query(models);
+  auto tasks = Exp4Policy::select_predict_tasks(state, query, 1000);
+  ASSERT_NE(model_3.second, tasks.front().model_.second);
+}
+
+TEST_F(Exp4Test, SerializationTest) {
+  auto bytes = Exp4Policy::serialize_state(state);
+  auto new_state = Exp4Policy::deserialize_state(bytes);
+  ASSERT_EQ(state.weight_sum_, new_state.weight_sum_);
+}
+
+// *****************
+// * EpsilonGreedy *
+// *****************
+class EpsilonGreedyTest : public ::testing::Test {
+ public:
+  virtual void SetUp() {
+    models.emplace_back(model_1);  // good
+    models.emplace_back(model_2);  // so-so
+    models.emplace_back(model_3);  // bad
+    state = EpsilonGreedyPolicy::initialize(models);
+  }
+  std::vector<VersionedModelId> models;
+  VersionedModelId model_1 = std::make_pair("classification", 0);
+  VersionedModelId model_2 = std::make_pair("regression", 1);
+  VersionedModelId model_3 = std::make_pair("random_forest", 2);
+  int times = 1000;
+  PolicyState state;
+};
+
+TEST_F(EpsilonGreedyTest, UpdateTest) {
+
+  auto feedback = Utility::create_feedback(20);
+  std::vector<Output> predictions;
+  while (times > 0) {
+    auto y_hat = rand() % 100;
+    if (y_hat < 33) {
+      predictions = Utility::create_predictions(model_1, y_hat);
+    } else if (y_hat > 66) {
+      predictions = Utility::create_predictions(model_3, y_hat);
+    } else {
+      predictions = Utility::create_predictions(model_2, y_hat);
+    }
+    state = EpsilonGreedyPolicy::process_feedback(state, feedback, predictions);
+    times -= 1;
+  }
+  ASSERT_GT(state.model_map_[model_3]["expected_loss"],
+            state.model_map_[model_2]["expected_loss"]);
+  ASSERT_GT(state.model_map_[model_2]["expected_loss"],
+            state.model_map_[model_1]["expected_loss"]);
+}
+
+TEST_F(EpsilonGreedyTest, SerializationTest) {
+  auto bytes = EpsilonGreedyPolicy::serialize_state(state);
+  auto new_state = EpsilonGreedyPolicy::deserialize_state(bytes);
+  ASSERT_EQ(state.weight_sum_, new_state.weight_sum_);
+}
```

Comment:
I don't know how to test Epsilon Greedy Selection cuz there is randomness in it so I can't assert anything. do you have any suggestions?